### ISSUE TITLE
RPN loss defined differently from the official setup

### ISSUE THREAD

**Issue by AKMourato** [facebookresearch/detectron2#3505]
Greetings.
I notice that the objectness loss and localization loss terms are normalized following:
https://github.com/facebookresearch/detectron2/blob/5e2a1ecccd228227c5a605c0a98d58e1b2db3640/detectron2/modeling/proposal_generator/rpn_outputs.py#L397-L399

In the Mask R-CNN paper, it is stated that all the RPN computations and hyperparameters are identical to the Faster R-CNN. However, Faster R-CNN paper weighs the losses as:

> L_rpn =  1/(N_cls)* L_cls + \lambda * 1/N_loc * L_loc

meaning: objectness loss over the number of re-sampled boxes (equal to RPN_batch_size) and localization loss times lambda (=10 by default) over the number of positives anchor boxes (foreground). 
Which differ from the normalization stated above done, as RPN_batch_size(256 by default) * batch size (nºimages).

@ppwwyyxx, what is the reason behind this?

**Comment by sukjunhwang** [comment]
The number of foregrounds may highly vary by the sampled training image, so probably it shares the same motivation as below.
https://github.com/facebookresearch/detectron2/blob/5e2a1ecccd228227c5a605c0a98d58e1b2db3640/detectron2/modeling/roi_heads/fast_rcnn.py#L249-L261

To note, there are many configurations which are different from original papers, but heuristically proven better.
Hope this answers your question.

**Comment by AKMourato** [comment]
Well, those comments make a lot of sense. Thank you @sukjunhwang, it shall be for that reason certainly, not sure if only but that seems enough.

Yes, so I gather. It was easier if every different sub-optimal config was indicated somewhere.
Cheers
